### ISSUE TITLE
*: enhance the tests for in-txn queries in TiDB

### ISSUE THREAD

**Issue by bb7133** [pingcap/tidb#18836]
## Development Task
There are some existing frameworks/tools designed for the correctness test against SQL-based databases, and some of them have been used in TiDB, like:

* [SQLSmith](https://github.com/pingcap/tipocket/tree/master/pkg/go-sqlsmith)
* [go-randgen](https://github.com/pingcap/go-randgen)

However, those tools are focused on the generating of query statements.

In TiDB, sometimes things are further than those 'pure queries' since there is a local transaction buffer in TiDB-server, for queries inside an on-the-fly transaction, data is retrieved from both the storage and the buffer. What's more, for the consideration of performance, some optimizations are made when data is retrieved from the transaction buffer.

So it would be very useful if we can adjust those existing test tools and validate the results for the queries inside a transaction.

For example, a simple routine of a query-based test may looks like:

```
For some random rules or iteration of a given AST:
  Generate query statement
  Run and validate the result
  ...
```

We can also try:
```
START TRANSACTION;
For some random rules or iteration of a given AST:
  Generate query statement
  Run and validate the result
  ...
COMMIT/ROLLBACK
```

By the way above the tests for in-transaction queries will be enhanced.

**Comment by gregwebs** [comment]
This is related to #18814

**Comment by shuke987** [comment]
We have developed a demo at https://github.com/zyguan/go-randgen/tree/tp-test/tp-test. It uses randgen to generate transaction test cases. We will add more randgen yy files to cover more scenes.

**Comment by cfzjywxk** [comment]
With the usage of the [tp-test](https://github.com/zyguan/go-randgen/tree/tp-test/tp-test), combining test cases for the OLTP scenarios, we've found several bugs with "critical" severity for the SQL execution and transaction layer, these are the [issues](https://github.com/pingcap/tidb/issues?q=is%3Aissue+author%3Azyguan+label%3Aseverity%2Fcritical), most of them are fixed in the latest TiDB release version.

**Comment by shuke987** [comment]
All these bugs have been fix in v4.0.9, and this tool has been added to release testing. We will keep enhancing and running these tests in the background. Issue closed.
